cassiopeia/solver/distance_correction/_crispr_cas9_distance_correction_solver.py

```diff
@@ -0,0 +1,556 @@
+            s1=s1,
+            s2=s2,
+            missing_state_indicator=missing_state_indicator,
+        )
+
+
+class CRISPRCas9DistanceCorrectionSolver(CassiopeiaSolver):
+    """
+    Apply a distance solver to corrected CRISPR-Cas9 distances.
+
+    Here, "corrected distances" are estimates of true pairwise tree distance,
+    as opposed to, say, Hamming distances.
+
+    In the distance correction scheme, we first estimate the two parameters of
+    the CRISPR-Cas9 model (the mutation rate and the collision probability),
+    and then use these values to compute corrected distances from the character
+    states of each leaf.
+
+    Assumes that the missing data indicator is negative, and that all mutated
+    states are positive, for simplicity.
+
+    Technical note: The reason why we use a "distance_corrector_name" instead
+    of just injecting the distance_corrector_function directly is because the
+    latter approach fails due to Numba compilation issues.
+
+    Args:
+        mutation_proportion_estimator: The mutation proportion estimator.
+            Default is `crispr_cas9_default_mutation_proportion_estimator`,
+            which uses the simple estimate #alleles>0 / #alleles>=0.
+            The mutation proportion can be set up front (i.e. hardcoded)
+            with the `crispr_cas9_hardcoded_mutation_proportion_estimator`.
+        collision_probability_estimator: The collision probability estimator.
+            Default is `crispr_cas9_default_collision_probability_estimator`,
+            which uses the simple estimate sum q_i^2, where each q_i is
+            estimated as q_i = #alleles=i / #alleles>0. The collision
+            probability can be set up front (i.e. hardcoded) with the
+            `crispr_cas9_hardcoded_collision_probability_estimator`.
+        distance_corrector_name: The name of the function that computes
+            corrected distances given the mutation rate, collision probability,
+            and character states of the two leaves. Allowed options:
+            `crispr_cas9_corrected_ternary_hamming_distance` (default) and
+            `crispr_cas9_corrected_hamming_distance`. Generally, the ternary
+            Hamming distance tends to perform better.
+        distance_solver: The distance solver used to reconstruct the tree
+            topology using the corrected distances. By default, NJ with rooting
+            is used, but other distance solvers such as UPGMA can be used too.
+    """
+
+    def __init__(
+        self,
+        mutation_proportion_estimator: MutationProportionEstimatorType = crispr_cas9_default_mutation_proportion_estimator,  # noqa
+        collision_probability_estimator: CollisionProbabilityEstimatorType = crispr_cas9_default_collision_probability_estimator,  # noqa
+        distance_corrector_name: str = "crispr_cas9_corrected_ternary_hamming_distance",  # noqa
+        distance_solver: DistanceSolver = NeighborJoiningSolver(add_root=True),
```

Comment:
All of these lines can be formatted to be <80 characters long I believe.

Reply:
I had been formatting my code with `isort code.py && black --line-length 80 code.py`, and if I added line breaks to these arguments `black` formated them back to one lone line each. But now I flanked the code with `"# fmt: off"` and `"# fmt: on"` so that `black` won't do this. Now all lines should be <= 80, with the caveat that `"# fmt: off"` and `"# fmt: on"` are there.